man/anoubis_notify.3
.\"	$OpenBSD: mdoc.template,v 1.9 2004/07/02 10:36:57 jmc Exp $
.\"
.\" Copyright (c) 2007 GeNUA mbH <info@genua.de>
.\"
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd April 15, 2008
.Dt ANOUBIS_NOTIFY 3
.Os Anoubis
.Sh NAME
.Nm struct anoubis_notify_group
.Nm struct anoubis_notify_head
.Nm anoubis_notify_callback_t
.Nm anoubis_notify_create
.Nm anoubis_notify_destroy
.Nm anoubis_notify_register
.Nm anoubis_notify_unregister
.Nm anoubis_notify_create_head
.Nm anoubis_notify_destroy_head
.Nm anoubis_notify
.Nm anoubis_notify_answer
.Nm anoubis_notify_sendreply
.Sh SYNOPSIS
.In anoubis_msg.h
.In anoubis_notify.h
.Ft struct anoubis_notify_group *
.Fn anoubis_notify_create "struct achat_channel * chan" "uid_t uid"
.Ft void
.Fn anoubis_notify_destroy "struct anoubis_notify_group *"
.Ft int
.Fn anoubis_notify_register "struct anoubis_notify_group * g" "uid_t uid" "u_int32_t ruleid" "u_int32_t subsystem"
.Ft int
.Fn anoubis_notify_unregister "struct anoubis_notify_group * g" "uid_t uid" "u_int32_t ruleid" "u_int32_t subsystem"
.Ft void
.Fn "(*anoubis_notify_callback_t)" "struct anoubis_notify_head * head" "int verdict" "void * cbdata"
.Ft struct anoubis_notify_head *
.Fn anoubis_notify_create_head "struct anoubis_msg * m" "anoubis_notify_callback_t finish" "void * cbdata"
.Ft void
.Fn anoubis_notify_destroy_head "struct anoubis_notify_head * h"
.Ft int
.Fn anoubis_notify "struct anoubis_notify_group * ng" "struct anoubis_notify_head * head" "unsigned int limit"
.Ft int
.Fn anoubis_notify_answer "struct anoubis_notify_group * ng" "anoubis_token_t token" "int verdict" "int delegate"
.Ft int
.Fn anoubis_notify_sendreply "struct anoubis_notify_head * head" "int verdict" "void * you" "uid_t uid"
.Sh DESCRIPTION
The
.Nm anoubis_notify.h
file provides facilities for handling notify messages in the anoubis
server.
There are two kinds of handles that are created by this module:
.Bl -bullet
.It
.Nm anoubis_notify_group
encapsulates communication with a single dedicated peer.
.It
.Nm anoubis_notify_head
tracks a single notify event that might be sent to multiple notify groups.
.El
.Pp
Heads and groups are linked together by an internal data structure
called
.Nm anoubis_notify_event.
.Pp
A notify group will only receive notify events that it actually registered
for with
.Nm anoubis_notify_register.
An active registration can be revoked with
.Nm anoubis_notify_unregister.
.Sh LIFE CYCLE OF EVENTS
If a new event needs an answer the first thing to do is to
create a new
.Nm anoubis_notify_head.
All information about the event and any potential callback functions
must be provided when creating the head.
Among other things this means that the same message token is used
for all notify groups that receive the event.
.Pp
The next step is to send the event to one or more notify groups.
This is done by
.Nm anoubis_notify.
This function will return a positive value if the event could actually
be sent and zero if the event did not match any active registration of
the group.
The parameter
.Nm limit
limits the total number of events in the session that are waiting for
a reply.
If the limit is reached an error is returned.
.Pp
If the event is a notify and not an ask event (i.e. no reply from the
peer is required) this is all that must be done and the caller should
destroy the notify head immediately after the final call to
.Nm anoubis_notify.
Likewise if the message did not match any registration the notify head
must be destroyed by the caller right now because no further events will
happen.
.Pp
The remainder of this section only applies to ask events that actually
need an answer.
.Pp
Such a message requires two more messages per peer:
.Bl -bullet
.It
A message from the peer that provides a return code for the event.
This message can either indicate that the return code provided in the event
should be used unconditionally or that the peer wishes to delegate the
decision.
In the latter case the return code is on only used if all peers
delegate the event and the current delegation is the last one.
.It
A message to the peer that informs the peer of the actual return code.
.El
.Pp
Both messages must be exchanged with
.Em each peer
that originally received the ask event.
If multiple peers try to answer the event unconditionally the first one
wins and the reply from the other peer is treated as if it were a delegation.
A peer that has not yet sent a return code or a delegation for an event
must send a delegation immediately if it receives a message with the
actual return code.
.Pp
This way the order in which the two messages described above are
sent does not matter.
.Pp
If a result message from a peer is received it must be processed
by calling
.Nm anoubis_notify_answer
on the appropriate group.
This is normally done by the server side protocol handler.
If the server wants to abort an event without an answer from any client
it can do so by calling
.Nm anoubis_notify_sendreply.
However, this function is normally called internally as a reaction to
a call to
.Nm anoubis_notify_answer.
.Pp
If a callback function for an event has been provided it will be called
exactly once for each notify head when a return code has been received.
The head itself can safely be destroyed from the callback function.
.Pp
Destroying a notify group will delegate all pending events without actually
providing a return code, i.e. the default return code will be used.
.Sh FUNCTION REFERENCE
.Bl -tag -width "MMM"
.It anoubis_notify_create
Create a notify group for the given channel.
Channels cannot be shared among multiple notify groups.
.It anoubis_notify_destroy
Destroy a notify group.
This will delegate all pending events of that group.
.It anoubis_notify_register
.It anoubis_notify_unregister
Register or unregister for a set of events.
Each event is sent only once even if it matches multiple registrations.
Registrations are cumulative and unregistration is only possible with the
exact same parameters as a previous registration.
.It anoubis_notify_create_head
Create a new notify head.
The caller must provide the event in the form of an
.Xr anoubis_msg 3
with a type of
.Em ANOUBIS_N_ASK
for ASK events that need an answer or
.Em ANOUBIS_N_NOTIFY
or
.Em ANOUBIS_N_LOGNOTIFY
for notifications or
.Em ANOUBIS_N_POLICYCHANGE
for policy change messages or
.Em ANOUBIS_N_STATUSNOTIFY
for generic status change messages.
If the function returns non-NULL the message will be destroyed together
with the head and caller must no longer use this message.
In case of an error, i.e. NULL is returned, the caller must free the message.
.It anoubis_notify
Send the notification described by the notify head to the given notify group.
.It anoubis_notify_answer
Process a return code for the message given by token.
The parameter delegate should be true if the peer wants to delegate
the decision.
.It anoubis_notify_sendreply
Send replies to all peers without having received an answer from any of them.
The value of the parameter
.Va you
should be
.Em NULL
if this function is called from extern.
.El
.Sh RETURN VALUES
.Nm anoubis_notify_create
and
.Nm anoubis_notify_create_head
return the new notify group or head respectively.
If the creation fails NULL is returned.
.Pp
.Nm anoubis_notify_register,
.Nm anoubis_notify_unregister,
.Nm anoubis_notify_answer
and
.Nm anoubis_notify_sendreply
return zero on success and a negative errno value in case of an error.
.Pp
.Nm anoubis_notify
returns a negative errno value in case of an error, zero if the message
was not sent because it did not match any registration and one if the
message was successfully sent.
.Sh SEE ALSO
.Xr anoubis_client 3 ,
.Xr anoubis_server 3 ,
.Xr anoubis_msg 3 ,
.Xr anoubis_chat 3 ,
.Xr errno 2 ,
.Sh AUTHORS
Christian Ehrhardt <ehrhardt@genua.de>


ChangeLog
2010-03-18 - Changes between 0.9.2 and 0.9.3

New features:

o anoubisd: Remove unnecessary fsync call in anoubisd/sfs.c
  which resulted in a huge performance loss when adding checksums.
o anoubisd: Now uses the multi-checksum-request messages to improve
  performance.
o xanoubis: Now uses the multi-checksum-request messages to improve
  performance.
o anoubisd: Limit the size of policy-files which can be configured in
  anoubisd.conf(5).
o xanoubis: Implementation of client-authorization was added to xanoubis.
o xanoubis: Support for editing default policies.
o xanoubis: Provide a dialogue to generate key-pair within the GUI.
o xanoubis: comprehensive performance improvements regarding the handling
  of very large policies within xanoubis.
o xanoubis: The nosfs-flag can now be set within the RuleEditor and
  RuleWizard.
o Support for SCTP was added to apn, anoubis daemon and xanoubis.
o Support for eCryptfs was added to the Anoubis Linux kernel.
o Authentication was added in protocol version 4.
o A multi-checksum-request message type was added in protocol version 5.
o Long-running jobs comprised of numerous tasks such as register and validate
  operations can now be interrupted within xanoubis.

Bug fixes:

o Linux Kernel: Fix a NULL pointer dereference sfs_bprm_committed_creds gets
  called with a task that does not have a security label attached to it.
  BUG #1447
o OpenBSD: Fix OpenBSD pool allocator bug.
o OpenBSD: anoubisctl leaks the setgid status to anoubisd causing issues
  on opening /dev/eventdev
o OpenBSD kernel: Fix an OpenBSD bug that can cause crashes in namei NFS code.
o OpenBSD kernel: Fix deadlocks with vnode locks in OpenBSD.
o Kernel: Disable ANOUBIS_REQUEST_STATS ioctl for regular users.
o anoubisd: Fix memory leaks in the anoubis daemon revealed by valgrind(1)
o Fix a crash in the SHA256 calculation of libanoubissig.
o xanoubis: Fix a display bug that caused text to be displayed in adjacent
  empty cells.
o xanoubis: Assure that the private key is loaded only into memory when actually
  needed.
o xanoubis: Fix timeout of end-of-session key within SFS-Browser.
o xanoubis: Fix a bug were index counting for visible/invisible columns could
  lead to a memory access behind the actual end of the list in question.
o xanoubis: Fix a segfault when xanoubis gets closed while in the upgrade
  message dialogue.
o xanoubis: Fix view in RuleEditor to show non-bold fonts in headers, refuse
selection of these and store the size persistently.
o xanoubis: Change permissions of .xanoubis directory to 0700.
o xanoubis: Fix AnPickFromFS Dialogue to handle files as well as directories.
o xanoubis: Fix a segfault if a PolicyCtrl object is destroyed without an
  explicit creation.
o xanoubis: Handling of the TrayIcon exclamation mark is now correctly done
  when opening LogViewer or Notification tab
o anoubis-keyinstall: Bail out and provide and exit with a reasonable error
  message when the tool is called by non-root users.
o Properly handle syssigs on !SFS_CS_OK filesystems to allow execs on dazukofs
  mounts.

Other changes:

o Limit connections per user to avoid DoS by ENFILE
o anoubisd: Enforce secure exec if the policy of a task changes during exec.
o anoubisd: The number of pending notifications per session was limited to
  avoid an out of memory DOS.
o xanoubis: Fix ComCsumAddTask to calculate checksum/signature in CsumCalcTask
  not until both are meant to be send to daemon to improve performance.
o xanoubis: Redesign of the ALF, SB and SFS overview module.
o xanoubis: Double-Clicking on an empty application rules now also opens the
  RuleEditor.
o xanoubis: A dialogue is shown to the user if a non-matching key-pair was
  detected.
o xanoubis: Redesign of the RuleEditor which now uses a grid widget.
o xanoubis: Provide better error messages if an error occurs during
  authentication.
o Linux kernel: Update to 2.6.32
o Initial implementation of a buffer overflow protection library used by
  anoubisd
o Fix quoting in APN to handle filenames containing quotes properly.
o Add README.privsep as extra dist file providing information on privilege
  separation.


2009-11-03 - Changes between 0.9.1 and 0.9.2

New features:

  o Initial support for dazukoFS: Make sure anoubisd is unaffected by dazukoFS
    and include dazukoFS in the kernel packages.
  o Handle OS upgrades correctly even if anoubisd is restarted in the process.
    This is done by delaying the restart until after the upgrade finished.
  o Anoubisctl/sfssig/xanoubis will now also log errors to the syslog that may
    be interesting for the administrator.

Bug fixes:

  o OpenBSD kernel: Fix an OpenBSD bug that can cause crashes in nfs_reply.
  o xanoubis: Fix directory path interfering with filter/invert options.
    BUG #1168
  o anoubisd/anoubisctl: Do not leak pid filehandle to childs.
  o xanoubis: Various fixes for operating at low screen resolutions. BUG #893
  o sfssig: Fix exporting of signatures.
  o anoubisd: Fix /etc/services not being available to anoubisd childs.
  o xanoubis: Fix 'Unexpected return-code: 1' error message. BUG #1373
  o xanoubis: Fix the Status field in the rule editor being updated too
    late. BUG #1369
  o xanoubis: Display correct state in Upgrade view.
  o xanoubis: Don't try to create a rule for an unknown binary after an
    ask event.
  o xanoubis: Make large SFS error messages scrollable. BUG #1352
  o Linux kernel: Fix EBADF when writing to a raw socket. This sometimes
    prevented dhclient from working.
  o sfssig: Fix 'add' failing if file is open for writing. BUG #1342
  o Linux kernel: Make sure a new file is not created if write access was
    denied.

Other changes:

  o Send upgrade-notify events to all attached UIs. Show the upgrade dialogue
    in xanoubis when such an event is received. Also use this to get rid of
    INotify.
  o xanoubis: Give a general overview message for SFS errors.
  o xanoubis: Save table column widths in xanoubis.conf.

src/libanoubis/anoubisprotocol/server/anoubis_notify.c
/*
 * Copyright (c) 2008 GeNUA mbH <info@genua.de>
 *
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
 * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
 * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
 * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
 * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
#include <config.h>
#include <stdlib.h>
#include <sys/types.h>
#include <anoubis_chat.h>
#include <anoubis_msg.h>
#include <anoubis_notify.h>
#include <anoubis_errno.h>

#include <sys/queue.h>

#ifdef LINUX
#include <linux/anoubis.h>
#endif
#ifdef OPENBSD
#include <dev/anoubis.h>
#endif

struct anoubis_notify_reg {
	LIST_ENTRY(anoubis_notify_reg) next;
	u_int32_t ruleid;
	uid_t uid;
	u_int32_t subsystem;
};

struct anoubis_notify_head {
	LIST_HEAD(, anoubis_notify_event) events;
	struct anoubis_msg * m;
	int verdict;
	int eventcount;
	anoubis_notify_callback_t finish;
	void * cbdata;
	void * you;
	uid_t uid;
	int reply;
};

/*
 * FLAG_VERDICT: This peer will not send a verdict (because it
 *    already did so, because it died or because it delegated the event)
 * FLAG_REPLY: A reply has been sent to this peer and we are disconnected
 *    from the head.
 */

#define FLAG_VERDICT	1
#define FLAG_REPLY	2
#define DROPMASK	(FLAG_VERDICT|FLAG_REPLY)

struct anoubis_notify_event {
	LIST_ENTRY(anoubis_notify_event) next;
	LIST_ENTRY(anoubis_notify_event) nextgroup;
	struct anoubis_notify_group * grp;
	struct anoubis_notify_head * head;
	unsigned int flags;
	anoubis_token_t token;
};

struct anoubis_notify_group {
	uid_t uid;	/* Authorized User-ID. */
	LIST_HEAD(, anoubis_notify_reg) regs;
	/*@dependent@*/
	struct achat_channel * chan;
	LIST_HEAD(, anoubis_notify_event) pending;
};

/*
 * The following code utilizes list functions from BSD queue.h, which
 * cannot be reliably annotated. We therefore exclude the following
 * functions from memory checking.
 */
/*@-memchecks@*/
struct anoubis_notify_group * anoubis_notify_create(struct achat_channel * chan,
    uid_t uid)
{
	struct anoubis_notify_group * ret;

	if (NULL == chan)
		return NULL;

	ret = malloc(sizeof(*ret));
	if (!ret)
		return NULL;

	ret->chan = chan;
	ret->uid = uid;
	LIST_INIT(&ret->regs);
	LIST_INIT(&ret->pending);
	return ret;
}

static void drop_event(struct anoubis_notify_event * ev, int flags);

void anoubis_notify_destroy(struct anoubis_notify_group * ng)
{
	struct anoubis_notify_reg * reg;

	while(!LIST_EMPTY(&ng->regs)) {
		reg = LIST_FIRST(&ng->regs);
		LIST_REMOVE(reg, next);
		free(reg);
	}
	while(!LIST_EMPTY(&ng->pending)) {
		struct anoubis_notify_event * nev;
		struct anoubis_notify_head * head;
		nev = LIST_FIRST(&ng->pending);
		head = nev->head;
		drop_event(nev, FLAG_VERDICT|FLAG_REPLY);
		if (!head || head->eventcount)
			continue;
		/*
		 * We just dropped the last event that might have
		 * answered this without getting an answer.
		 */
		anoubis_notify_sendreply(head, head->verdict, head->you,
		    head->uid);
	}
	free(ng);
}

/*
 * Drop pending events and free the head. When this function is called
 * the list should be empty. We still drop the events to be sure.
 */
void anoubis_notify_destroy_head(struct anoubis_notify_head * head)
{
	anoubis_msg_free(head->m);
	while(!LIST_EMPTY(&head->events)) {
		struct anoubis_notify_event * event;
		event = LIST_FIRST(&head->events);
		drop_event(event, FLAG_REPLY);
	}
	free(head);
}

static int anoubis_register_ok(struct anoubis_notify_group * ng,
    uid_t uid, u_int32_t ruleid, u_int32_t subsystem)
{
	/* Registering for stat messages is allowed if all other ids are 0 */
	if (subsystem == ANOUBIS_SOURCE_STAT)
		return (uid == 0 && ruleid == 0);
	/*
	 * Registering for other events is only allowed if the uid and
	 * the authorized uid match or the user is root.
	 * uid -1 is a special case, which means that (unfortunately)
	 * this user can't monitor events.
	 */
	if (ng->uid == 0 /* XXX root */ ||
	   ((ng->uid == uid) && (ng->uid != (uid_t) -1)))
		return 1;
	return 0;
}

int anoubis_notify_register(struct anoubis_notify_group * ng,
    uid_t uid, u_int32_t ruleid, u_int32_t subsystem)
{
	struct anoubis_notify_reg * reg;

	if (!anoubis_register_ok(ng, uid, ruleid, subsystem))
		return -EPERM;
	reg = malloc(sizeof(struct anoubis_notify_reg));
	if (!reg)
		return -ENOMEM;
	reg->uid = uid;
	reg->ruleid = ruleid;
	reg->subsystem = subsystem;
	LIST_INSERT_HEAD(&ng->regs, reg, next);
	return 0;
}

static int reg_match(struct anoubis_notify_reg * reg,
    uid_t uid, u_int32_t ruleid, u_int32_t subsystem)
{
	/*
	 * XXX root can use uid -1 to request all notifications
	 */
	return ((reg->uid == (uid_t) -1 || reg->uid == uid)
	    && (!reg->ruleid || reg->ruleid == ruleid)
	    && (!reg->subsystem || reg->subsystem == subsystem));
}

static int reg_match_all(struct anoubis_notify_group * ng,
    uid_t uid, u_int32_t ruleid, u_int32_t subsystem)
{
	struct anoubis_notify_reg * reg;
	LIST_FOREACH(reg, &ng->regs, next) {
		 if (reg_match(reg, uid, ruleid, subsystem))
			return 1;
	}
	return 0;
}

static void drop_event(struct anoubis_notify_event * ev, int flags)
{
	if ((flags & FLAG_VERDICT) && (ev->flags & FLAG_VERDICT) == 0) {
		ev->flags |= FLAG_VERDICT;
		/*
		 * No need to track the eventcount if a reply has already
		 * been sent. Moreover, ev->head is NULL in this case anyway.
		 */
		if ((ev->flags & FLAG_REPLY) == 0)
			ev->head->eventcount--;
	}
	if ((flags & FLAG_REPLY) && (ev->flags & FLAG_REPLY) == 0) {
		LIST_REMOVE(ev, nextgroup);
		ev->head = NULL;
		ev->flags |= FLAG_REPLY;
	}
	if ((ev->flags & DROPMASK) == DROPMASK) {
		LIST_REMOVE(ev, next);
		free(ev);
	}
}


int anoubis_notify_unregister(struct anoubis_notify_group * ng,
    uid_t uid, u_int32_t ruleid, u_int32_t subsystem)
{
	struct anoubis_notify_reg * reg;
	LIST_FOREACH(reg, &ng->regs, next) {
		 if (reg_match(reg, uid, ruleid, subsystem)) {
			LIST_REMOVE(reg, next);
			free(reg);
			return 0;
		 }
	}
	return -ESRCH;
}

struct anoubis_notify_head *
anoubis_notify_create_head(struct anoubis_msg * m,
    anoubis_notify_callback_t finish, void * cbdata)
{
	struct anoubis_notify_head * head;
	int opcode;

	opcode = get_value(m->u.general->type);
	switch(opcode) {
	case ANOUBIS_N_POLICYCHANGE:
		if (!VERIFY_LENGTH(m, sizeof(Anoubis_PolicyChangeMessage)))
			return NULL;
		break;
	case ANOUBIS_N_STATUSNOTIFY:
		if (!VERIFY_LENGTH(m, sizeof(Anoubis_StatusNotifyMessage)))
			return NULL;
		break;
	case ANOUBIS_N_NOTIFY:
	case ANOUBIS_N_ASK:
	case ANOUBIS_N_LOGNOTIFY:
		if (!VERIFY_LENGTH(m, sizeof(Anoubis_NotifyMessage)))
			return NULL;
		if (opcode != ANOUBIS_N_LOGNOTIFY) {
			set_value(m->u.notify->error, 0);
			set_value(m->u.notify->loglevel, 0);
		}
		break;
	default:
		return NULL;
	}
	head = malloc(sizeof(struct anoubis_notify_head));
	if (!head)
		return NULL;
	LIST_INIT(&head->events);
	head->m = m;
	head->verdict = ANOUBIS_E_IO;
	head->eventcount = 0;
	head->finish = finish;
	head->cbdata = cbdata;
	head->reply = 0;
	head->uid = 0;
	head->you = NULL;
	return head;
}

/*
 * Returns:
 *  - negative errno on error.
 *  - zero if session is not registered for the message
 *  - positive if notify message was sent.
 */

int anoubis_notify(struct anoubis_notify_group * ng,
    struct anoubis_notify_head * head, unsigned int limit)
{
	struct anoubis_notify_event * nev;
	struct anoubis_msg * m = head->m;
	int ret, opcode;
	u_int32_t uid, ruleid, subsystem;
	anoubis_token_t token;
	unsigned int pendingcnt = 0;

	opcode = get_value(m->u.general->type);
	switch(opcode) {
	case ANOUBIS_N_POLICYCHANGE:
		uid = get_value(m->u.policychange->uid);
		subsystem = ANOUBIS_SOURCE_STAT;
		ruleid = 0;
		token = 0;
		/*
		 * Only root receives policy change messages for other
		 * users. XXX CEH: This check should be somewhere else!
		 */
		if (ng->uid != 0 && ng->uid != uid)
			return 0;
		break;
	case ANOUBIS_N_STATUSNOTIFY:
		subsystem = ANOUBIS_SOURCE_STAT;
		uid = 0;
		ruleid = 0;
		token = 0;
		break;
	default:
		uid = get_value(m->u.notify->uid);
		ruleid = get_value(m->u.notify->rule_id);
		subsystem = get_value(m->u.notify->subsystem);
		token = m->u.notify->token;
		if (token == 0)
			return -EINVAL;
	}

	if (!reg_match_all(ng, uid, ruleid, subsystem))
		return 0;
	LIST_FOREACH(nev, &ng->pending, next) {
		pendingcnt++;
		if (nev->token == token)
			return -EEXIST;
	}
	/* In these cases there is no need to wait for a reply */
	if (opcode == ANOUBIS_N_NOTIFY || opcode == ANOUBIS_N_LOGNOTIFY
	    || opcode == ANOUBIS_N_POLICYCHANGE
	    || opcode == ANOUBIS_N_STATUSNOTIFY) {
		ret = anoubis_msg_send(ng->chan, m);
		if (ret < 0)
			return ret;
		return 1;
	}
	if (limit && pendingcnt >= limit)
		return -ENOSPC;
	nev = malloc(sizeof(struct anoubis_notify_event));
	if (!nev)
		return -ENOMEM;
	nev->token = token;
	nev->flags = 0;
	nev->head = head;
	nev->grp = ng;
	ret = anoubis_msg_send(ng->chan, m);
	if (ret < 0) {
		free(nev);
		return ret;
	}
	LIST_INSERT_HEAD(&ng->pending, nev, next);
	LIST_INSERT_HEAD(&head->events, nev, nextgroup);
	head->eventcount++;
	return 1;
}

int anoubis_notify_sendreply(struct anoubis_notify_head * head,
    int verdict, void * you, uid_t uid)
{
	struct anoubis_msg * m;
	anoubis_notify_callback_t finish;
	void * cbdata;

	if (verdict < 0)
		return -EINVAL;
	if (head->reply)
		return 0;
	head->reply = 1;
	m = anoubis_msg_new(sizeof(Anoubis_NotifyResultMessage));
	if (!m)
		return -ENOMEM;
	head->verdict = verdict;
	head->uid = uid;
	set_value(m->u.notifyresult->uid, uid);
	set_value(m->u.notifyresult->error, verdict);
	while(!LIST_EMPTY(&head->events)) {
		int ret;
		struct anoubis_notify_event * nev;
		nev = LIST_FIRST(&head->events);
		if (you == nev)
			set_value(m->u.notifyresult->type, ANOUBIS_N_RESYOU);
		else
			set_value(m->u.notifyresult->type, ANOUBIS_N_RESOTHER);
		m->u.notifyresult->token = nev->token;
		ret = anoubis_msg_send(nev->grp->chan, m);
		if (ret < 0)
			drop_event(nev, FLAG_VERDICT);
		drop_event(nev, FLAG_REPLY);
	}
	anoubis_msg_free(m);

	/* Call the callback function. This might leagally free head! */
	finish = head->finish;
	cbdata = head->cbdata;
	if (finish)
		finish(head, head->verdict, cbdata);
	return 0;
}

int anoubis_notify_answer(struct anoubis_notify_group * ng,
    anoubis_token_t token, int verdict, int delegate)
{
	struct anoubis_notify_event * nev;
	struct anoubis_notify_head * head;

	LIST_FOREACH(nev, &ng->pending, next) {
		if (nev->token == token)
			break;
	}
	/*
	 * If we got no mesage with this token or if we already received an
	 * answer for this token from this channel, this is no longer
	 * allowed.
	 */
	if (!nev || (nev->flags & FLAG_VERDICT))
		return -ESRCH;
	/*
	 * Event has already been answered. This is just the reply from
	 * the peer. In this case we only drop the event. We must not access
	 * nev->head!
	 */
	if (nev->flags & FLAG_REPLY) {
		drop_event(nev, FLAG_VERDICT);
		return 0;
	}
	/*
	 * If this is a delegation we remove the event from its group but
	 * leave it in the head (if it is still there). If dropping the
	 * event removes the last event from head take this as a verict
	 * with -EIO.
	 */
	head = nev->head;
	drop_event(nev, FLAG_VERDICT);
	/*
	 * Return if this is a delegation and answers from other
	 * peers are still outstanding just remember the verdict
	 * and return.
	 */
	head->verdict = verdict;
	head->uid = nev->grp->uid;
	head->you = nev;
	if (delegate) {
		if (head->eventcount)
			return 0;
	}
	/*
	 * Either this is the first reply or the last delegation.
	 * In either case send the replies now. In case of a delegation
	 * use the verdict provided together with the delegation!
	 */
	anoubis_notify_sendreply(head, head->verdict, head->you, head->uid);
	return 0;
}
/*@=memchecks@*/
